src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,694 @@
+            let redundant_list = purged_files.purge(remains_capacity);
+            for (i, fd) in redundant_list.fds.iter().enumerate() {
+                let file_id = FileId {
+                    seq: redundant_list.first_seq + i as FileSeq,
+                    queue: self.queue,
+                };
+                FileCollection::clear_file(
+                    self.file_system.as_ref(),
+                    &self.paths,
+                    file_id,
+                    fd,
+                    false, /* is_stale */
+                )?;
+            }
+            // Meanwhile, the new purged files from `self.active_files` should be RENAME
+            // to stale files with `.raftlog.stale` suffix, to reduce the unnecessary
+            // recovery timecost when RESTART.
+            for (i, fd) in purged_files.fds.iter().enumerate() {
+                let file_id = FileId {
+                    seq: purged_files.first_seq + i as FileSeq,
+                    queue: self.queue,
+                };
+                let stale_file_id = FileId {
+                    seq: stale_files.span().1 + 1_u64,
+                    queue: self.queue,
+                };
+                let path = file_id.build_file_path(&self.paths[fd.path_id]);
+                let stale_path = stale_file_id.build_stale_file_path(&self.paths[fd.path_id]);
+                self.file_system.reuse(&path, &stale_path)?;
+                stale_files.push_back(
+                    stale_file_id.seq,
+                    FileWithFormat {
+                        handle: Arc::new(self.file_system.open(&stale_path)?),
+                        path_id: fd.path_id,
+                        format: fd.format,
+                    },
+                );
+            }
+        }
+        Ok(logical_purged_count)
+    }
+
+    /// Initialize current file collection by preparing several
+    /// stale files for later log recycling in advance.
+    ///
+    /// Attention, this function only makes sense when
+    /// `Config.enable-log-recycle` is true.
+    pub fn initialize(&mut self) -> Result<()> {
+        let now = Instant::now();
+        let stale_capacity = std::cmp::min(
+            self.capacity.saturating_sub(self.active_files.read().len()),
+            max_stale_log_count(),
+        );
+        // If `stale_capacity` > 0, we should prepare stale files for later
+        // log recycling in advance.
+        if stale_capacity > 0 {
+            // Concurrent prepraring will bring more time consumption on racing. So, we just
+            // introduce a serial processing for preparing progress.
+            let mut stale_files = self.stale_files.write();
+            let (prepare_first_seq, prepare_stale_files_count) = (
+                stale_files.span().1 + 1,
+                stale_capacity.saturating_sub(stale_files.len()),
+            );
+            if stale_files.len() == 0 {
+                stale_files.first_seq = 1;
+            }
```

Comment:
Why? Isn't it already updated in `push_back`?

Reply:
Redundant here.